### DISCUSSION TITLE
cy.stub matchMedia to set dark mode causes other (not stubbed) calls to fail

### DISCUSSION THREAD

**Discussion by monktana** [cypress-io/cypress#29881] [Questions and Help]
Hello, i have come across [this](https://github.com/cypress-io/cypress/issues/8056) issue on my google search to solve an issue i currently have with stubbing the matchMedia function.

I'm have a test case to ensure that my color mode toggle works the way i expect it to. Therefore, i want to (pre)set the dark mode in the test following this blogpost: [https://www.cypress.io/blog/2019/12/13/test-your-web-app-in-dark-mode](https://www.cypress.io/blog/2019/12/13/test-your-web-app-in-dark-mode)

```typescript
cy.visit('/', {
  onBeforeLoad(win: Cypress.AUTWindow) {
    cy.stub(win, 'matchMedia')
      .withArgs('(prefers-color-scheme: dark)')
      .as('set dark mode')
      .returns({
        matches: true,
        addListener: () => {},
      });
  },
});
```

When executing the test, i get this error message:

![image](https://github.com/user-attachments/assets/a117c6c1-5752-4f28-9598-9c6d2015713f)

Can someone tell me why this fails? Do i need to manually create another stub for no arguments (or the 'prefers-reduced-motion')?

What is the return of the automatically created stub?

Thanks in advance :)

**Comment by portyu9** [comment]
Hi @monktana 

The issue is that `cy.stub(win, 'matchMedia')` replaces **all** calls to `window.matchMedia()`.

Your `.withArgs('(prefers-color-scheme: dark)')` only defines the return value for that specific argument. Any other call, such as `prefers-reduced-motion`, falls back to the parent stub, which returns `undefined` by default. If your application then reads `.matches`, it will fail.

You don't need a separate stub for every query. Give `matchMedia` a default implementation and make only the dark-mode query return `true`:

```js
cy.visit('/', {
  onBeforeLoad(win) {
    cy.stub(win, 'matchMedia').callsFake((query) => ({
      matches: query === '(prefers-color-scheme: dark)',
      media: query,
      onchange: null,
      addListener: () => {},
      removeListener: () => {},
      addEventListener: () => {},
      removeEventListener: () => {},
      dispatchEvent: () => false,
    }))
  },
})
```

Now the dark-mode query returns `matches: true`, while other media queries return a valid `MediaQueryList`-like object with `matches: false`.

So the failure isn't caused by `withArgs()` itself—the problem is that replacing `matchMedia()` also replaces its normal behavior for every argument you didn't configure.

**Reply by portyu9** [reply]
@monktana Please mark this answer as accepted if it helped, thank you.
